python/format/src/graphs.py

```diff
@@ -0,0 +1,243 @@
+    """
+    properties = set()
+    for _, _, property in edges:
+        properties.add(property)
+    for property in mandatory_properties:
+        if property not in properties:
+            error = f'Property "{property}" is mandatory, but does not exist. Existing properties: {properties}.'
+            issues.add_error(error)
+    for property in optional_properties:
+        if property not in properties:
+            error = f'Property "{property}" is recommended, but does not exist. Existing properties: {properties}.'
+            issues.add_warning(error)
+    for possible_exclusive_properties in exclusive_properties:
+        if not _there_exists_at_least_one_property(
+            properties, possible_exclusive_properties
+        ):
+            error = f"At least one of these properties should be defined: {possible_exclusive_properties}. Existing properties: {properties}."
+            issues.add_error(error)
+
+
+def _find_children_objects(edges: Edges, property: rdflib.term.URIRef):
+    """Finds all children objects/nodes."""
+    return [
+        _object
+        for _, _object, _property in edges
+        if isinstance(_object, rdflib.term.BNode) and _property == property
+    ]
+
+
+def _find_single_object_by_property(
+    issues: errors.Issues, edges: Edges, property: rdflib.term.URIRef
+):
+    """Finds value where `[property]: value` in edges."""
+    possible_objects = [
+        _object for _, _object, _property in edges if _property == property
+    ]
+    if len(possible_objects) != 1:
+        issues.add_error(
+            f'Property "{property}" is mandatory, but does not exist. Existing properties: {edges}.'
+        )
+    return possible_objects[0]
+
+
+def _find_name(issues: errors.Issues, edges: Edges):
+    return _find_single_object_by_property(issues, edges, constants.SCHEMA_ORG_NAME)
+
+
+def _check_node_has_type(
+    issues: errors.Issues, edges: Edges, _type: rdflib.term.URIRef
+):
+    """Checks for the presence of `"@type": _type` in edges."""
+    for _, _object, _property in edges:
+        if _property == constants.RDF_TYPE:
+            if _object != _type:
+                issues.add_error(
+                    f'Node should have an attribute `"@type": "{_type}"`. Found `"@type": "{_object}"`.'
+                )
+            return
+    issues.add_error(f'Node should have an attribute `"@type": "{_type}"`.')
+
+
+def _find_entry_object(
+    issues: errors.Issues, graph: nx.MultiDiGraph
+) -> rdflib.term.BNode:
+    """Finds the source entry node without any parent."""
+    sources = [
+        node
+        for node, indegree in graph.in_degree(graph.nodes())
+        if indegree == 0 and isinstance(node, rdflib.term.BNode)
+    ]
+    if len(sources) != 1:
+        issues.add_error(f"Trying to define more than one dataset in the file.")
+    return sources[0]
+
+
+def check_metadata(
+    issues: errors.Issues,
+    edges: Edges,
+    dataset_name: str,
+):
```

Comment:
Can you add a docstring to those functions please?

Reply:
Done.